idgen/idgen.go

```diff
@@ -0,0 +1,147 @@
+package idgen
+
+import (
+	"fmt"
+	"math/bits"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+)
+
+// HydraIdentityGenerator is a shared balanced ID generator.
+var HydraIdentityGenerator = NewBalancedIdentityGenerator()
+
+// BalancedIdentityGenerator is a facility for generating IPFS identities (i.e. IPFS private keys),
+// whose corresponding DHT keys are highly balanced, compared to just generating random keys.
+// Balancing is accomplished using "the power of two choices" paradigm:
+// https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf
+//
+// New identities are generated by calling AddBalanced. BalancedIdentityGenerator remembers
+// generated identities, in order to ensure balance for future identities.
+// Generated identities can be removed using Remove.
+//
+// BalancedIdentityGenerator maintains the invariant that all identities, presently in its memory,
+// form an almost-perfectly balanced set.
+type BalancedIdentityGenerator struct {
+	sync.Mutex
+	xorTrie *XorTrie
+	count   int
+}
+
+// NewBalancedIdentityGenerator creates a new balanced identity generator.
+func NewBalancedIdentityGenerator() *BalancedIdentityGenerator {
+	return &BalancedIdentityGenerator{
+		xorTrie: NewXorTrie(),
+	}
+}
+
+// AddUnbalanced is used for testing purposes. It generates a purely random identity,
+// which is not balanced with respect to the existing identities in the generator.
+// The generated identity is stored in the generator's memory.
+func (bg *BalancedIdentityGenerator) AddUnbalanced() (crypto.PrivKey, error) {
+	bg.Lock()
+	defer bg.Unlock()
+	p0, t0, _, err0 := bg.genUniqueID()
+	if err0 != nil {
+		return nil, fmt.Errorf("generating unbalanced ID candidate, %w", err0)
+	}
+	bg.xorTrie.Insert(t0)
+	bg.count++
+	return p0, nil
+}
+
+// AddBalanced generates a random identity, which
+// is balanced with respect to the existing identities in the generator.
+// The generated identity is stored in the generator's memory.
+func (bg *BalancedIdentityGenerator) AddBalanced() (crypto.PrivKey, error) {
+	bg.Lock()
+	defer bg.Unlock()
+	p0, t0, d0, err0 := bg.genUniqueID()
+	if err0 != nil {
+		return nil, fmt.Errorf("generating first balanced ID candidate, %w", err0)
+	}
+	p1, t1, d1, err1 := bg.genUniqueID()
+	if err1 != nil {
+		return nil, fmt.Errorf("generating second balanced ID candidate, %w", err1)
+	}
+	if d0 < d1 {
+		bg.xorTrie.Insert(t0)
+		bg.count++
+		return p0, nil
+	} else {
+		bg.xorTrie.Insert(t1)
+		bg.count++
+		return p1, nil
+	}
+}
+
+func (bg *BalancedIdentityGenerator) genUniqueID() (privKey crypto.PrivKey, trieKey TrieKey, depth int, err error) {
```

Comment:
What's the reason for naming the return values here?

Reply:
The reason was so that the returned "int" value can have a name ("depth") for clarity.
Does PL follow a specific Golang styleguide?